core/src/services/hdfs/backend.rs

```diff
@@ -290,45 +289,40 @@ impl Access for HdfsBackend {
     }
 
     async fn write(&self, path: &str, op: OpWrite) -> Result<(RpWrite, Self::Writer)> {
-        let (target_path, tmp_path) = if let Some(atomic_write_dir) = &self.atomic_write_dir {
-            let target_path = build_rooted_abs_path(&self.root, path);
-            let tmp_path = build_rooted_abs_path(atomic_write_dir, &tmp_file_of(path));
-
-            // If the target file exists, we should append to the end of it directly.
-            if op.append() && self.client.metadata(&target_path).is_ok() {
-                (target_path, None)
-            } else {
-                (target_path, Some(tmp_path))
+        let target_path = build_rooted_abs_path(&self.root, path);
+        let target_exists = match self.client.metadata(&target_path) {
+            Ok(_) => true,
+            Err(err) => {
+                if err.kind() != io::ErrorKind::NotFound {
+                    return Err(new_std_io_error(err));
+                }
+                false
             }
-        } else {
-            let p = build_rooted_abs_path(&self.root, path);
-            (p, None)
         };
 
-        if let Err(err) = self.client.metadata(&target_path) {
-            // Early return if other error happened.
-            if err.kind() != io::ErrorKind::NotFound {
-                return Err(new_std_io_error(err));
+        let should_append = op.append() && target_exists;
+        let tmp_path = self.atomic_write_dir.as_ref().and_then(|atomic_write_dir| {
+            // If the target file exists, we should append to the end of it directly.
+            if should_append {
+                None
+            } else {
+                Some(build_rooted_abs_path(atomic_write_dir, &tmp_file_of(path)))
             }
+        });
 
+        if !target_exists {
             let parent = get_parent(&target_path);
-
             self.client.create_dir(parent).map_err(new_std_io_error)?;
-
-            let mut f = self
-                .client
-                .open_file()
-                .create(true)
-                .write(true)
-                .async_open(&target_path)
-                .await
+        } else if tmp_path.is_some() {
+            // we must delete the target_path, otherwise the rename_file operation will fail
+            self.client
+                .remove_file(&target_path)
```

Comment:
Hi, what if we remove files during `Write::close`? This way, users won't lose their existing files if the write fails.

Reply:
Oh, yes. you're right. I'll fix it